linera-service/src/cli_wrappers/wallet.rs

```diff
@@ -141,26 +145,82 @@ impl ClientWrapper {
-            ])
-            .args(["--send-timeout-ms", "500000"])
-            .args(["--recv-timeout-ms", "500000"])
-            .arg("--wait-for-outgoing-messages");
+        let mut command = self.command_binary().await?;
+        command.current_dir(self.path_provider.path()).env(
+            "RUST_LOG",
+            std::env::var("RUST_LOG").unwrap_or(String::from("linera=debug")),
+        );
+        for argument in self.command_arguments() {
+            command.arg(&*argument);
+        }
         Ok(command)
     }
 
+    /// Returns an iterator over the arguments that should be added to all command invocations.
+    fn command_arguments(&self) -> impl Iterator<Item = Cow<'_, str>> + '_ {
+        [
+            "--wallet".into(),
+            self.wallet.as_str().into(),
+            "--storage".into(),
+            self.storage.as_str().into(),
+            "--max-pending-message-bundles".into(),
+            self.max_pending_message_bundles.to_string().into(),
+            "--send-timeout-ms".into(),
+            "500000".into(),
+            "--recv-timeout-ms".into(),
+            "500000".into(),
+            "--wait-for-outgoing-messages".into(),
+        ]
+        .into_iter()
+    }
+
+    /// Returns the [`Command`] instance configured to run the appropriate binary.
+    ///
+    /// The path is resolved once and cached inside `self` for subsequent usages.
+    async fn command_binary(&self) -> Result<Command> {
+        match self.command_with_cached_binary_path() {
+            Some(command) => Ok(command),
+            None => {
+                let resolved_path = resolve_binary("linera", env!("CARGO_PKG_NAME")).await?;
+                let command = Command::new(&resolved_path);
+
+                self.set_cached_binary_path(resolved_path);
+
+                Ok(command)
+            }
+        }
+    }
+
+    /// Returns a [`Command`] instance configured with the cached `binary_path`, if available.
+    fn command_with_cached_binary_path(&self) -> Option<Command> {
+        let binary_path = self
+            .binary_path
+            .lock()
+            .expect("Threads should not panic while holding a lock to `binary_path`");
+
+        binary_path.as_ref().map(Command::new)
+    }
+
+    /// Sets the cached `binary_path` with the `new_binary_path`.
+    ///
+    /// # Panics
+    ///
+    /// If the cache is already set to a different value. In theory the two threads calling
+    /// `command_binary` can race and resolve the binary path twice, but they should always be the
+    /// same path.
+    fn set_cached_binary_path(&self, new_binary_path: PathBuf) {
+        let mut binary_path = self
+            .binary_path
+            .lock()
+            .expect("Threads should not panic while holding a lock to `binary_path`");
+
+        if binary_path.is_none() {
+            *binary_path = Some(new_binary_path);
+        } else {
+            assert_eq!(*binary_path, Some(new_binary_path));
+        }
+    }
+
```

Comment:
nit: These two internal functions are a bit trivial and used only once. Can we inline them? Also, it is acceptable to use `.unwrap()` for locks, especially in test code.

Reply:
I had initially made the three functions a single `async fn`, but the compiler then complained about the resulting `Future` not being `Sync` because it has a `MutexGuard`, even when I added manual `drop`s for the guard before an `await` point and in the start of every conditional execution bifurcation (!?!?). Creating separate synchronous functions was the way I found to make it compile (after a lot of trial and error :confused: ).